File: neet_code_all_cpp/sliding_window/minimum_operations_to_reduce_x_to_zero.cpp
#include <algorithm>
#include <iostream>
#include <tuple>
#include <vector>

// Kotlinの模範解答より
class SolutionAnsKotlin {
    public:
        // AC
        int minOperations(std::vector<int>& nums, int x) {
            auto [left, max, current] = std::tuple(0, -1, 0);
            int target = -x; 
            std::for_each(nums.begin(), nums.end(), [&](auto x) { target += x;});

            for(int right = 0; right < nums.size(); ++right) {
                current += nums[right];

                while(left <= right && current > target) {
                    current -= nums[left];
                    ++left;
                }

                if(current == target) {
                    max = std::max(max, right - left + 1);
                }
            }

            if(max == -1) {
                return -1;
            }

            return (int)nums.size() - max;
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, int>;
    Case case_1 = {{1, 1, 4, 2, 3}, 5};
    // => 2
    Case case_2 = {{5, 6, 7, 8, 9}, 4};
    // => -1
    Case case_3 = {{3, 2, 20, 1, 1, 3}, 10};
    // => 5

    SolutionAnsKotlin s_ans_kt;

    std::cout << s_ans_kt.minOperations(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans_kt.minOperations(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans_kt.minOperations(case_3.first, case_3.second) << std::endl;
}

File: neet_code_all/src/sliding_window/minimum_operations_to_reduce_x_to_zero.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn min_operations(nums: Vec<i32>, x: i32) -> i32 {
        let mut x = x;
        let n = nums.len();
        let mut prefix_sum = vec![0; n + 1];
        let mut postfix_sum = vec![0; n + 1];
        for i in 0..n {
            prefix_sum[i + 1] = prefix_sum[i] + nums[i];
        }
        for i in (0..n).rev() {
            postfix_sum[i] = postfix_sum[i + 1] + nums[i];
        }
        println!("prefix_sum: {:?}", prefix_sum);
        println!("postfix_sum: {:?}", postfix_sum);

        -1
    }
}

// Kotlinの模範解答より
struct SolutionAnsKotlin {}
impl SolutionAnsKotlin {
    // AC
    pub fn min_operations(nums: Vec<i32>, x: i32) -> i32 {
        let (mut left, mut max, mut current) = (0, -1_i32, 0);
        let target = nums.iter().sum::<i32>() - x;

        for right in 0..nums.len() {
            current += nums[right];

            while left <= right && current > target {
                current -= nums[left];
                left += 1;
            }

            if current == target {
                max = max.max(right as i32 - left as i32 + 1);
            }
        }

        if max == -1 {
            return -1;
        }

        nums.len() as i32 - max
    }
}

fn main() {
    let case_1 = (vec![1, 1, 4, 2, 3], 5);
    // => 2
    let case_2 = (vec![5, 6, 7, 8, 9], 4);
    // => -1
    let case_3 = (vec![3, 2, 20, 1, 1, 3], 10);
    // => 5

    println!(
        "case_1: {}",
        SolutionAnsKotlin::min_operations(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsKotlin::min_operations(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsKotlin::min_operations(case_3.0.clone(), case_3.1)
    );
}
